instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py

```diff
@@ -666,9 +667,9 @@ async def otel_receive():
             with self.tracer.start_as_current_span(
                 " ".join((server_span_name, scope["type"], "receive"))
             ) as receive_span:
-                if callable(self.client_request_hook):
-                    self.client_request_hook(receive_span, scope)
                 message = await receive()
+                if callable(self.client_request_hook):
+                    self.client_request_hook(receive_span, scope, message)
```

Comment:
Im confused as to why we are changing the order of this? Conceptually this was supposed to represent a hook that users can add that will be called BEFORE the receive call is made and right AFTER a span is created, such as in [requests](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L216) and [flask](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L437). It also follows the precedence of the [server request hook](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L617). This way users can implement logic like timing how long a request has taken.

Reply:
> Conceptually this was supposed to represent a hook that users can add that will be called BEFORE the receive call is made and right AFTER a span is created, such as in [requests](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L216) and [flask](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L437).

I believe that my change actually makes it more aligned with those:

- In Flask, request is _already read_ and available in [`flask.request`](https://flask.palletsprojects.com/en/3.0.x/api/#flask.request) by the moment the hook is called. Similarly, `client_request_hook` would get called right after the event has been read, but still _before it gets processed_ in the [`self.app(...)`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L632) call down the chain
- In `requests`, the `request` is _already prepared_ by the moment the request hook is called
- In Django, the request hook is called [at the very end](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py#L289), when `request` is fully prepared

> It also follows the precedence of the [server request hook](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L617). This way users can implement logic like timing how long a request has taken.

Server request hook gets called before [`self.app(...)`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L632) and hence, before `otel_receive` even gets called and before the hook/receive execution order gets to play its role. So, my change does not affect it at all, and I'm not sure if I understand your point well. Could you elaborate on that?